File: src/java/eu/sig/training/ch03/binarytree/BinaryTreeNode.java
package eu.sig.training.ch03.binarytree;

public class BinaryTreeNode<T extends Comparable<T>> {
    private final T value;
    private BinaryTreeNode<T> left;
    private BinaryTreeNode<T> right;

    public BinaryTreeNode(T value) {
        this.value = value;
    }

    public void insert(T value) {
        if (value.compareTo(this.value) < 0) {
            if (left != null) {
                left.insert(value);
            } else {
                left = new BinaryTreeNode<T>(value);
            }
        } else {
            if (right != null) {
                right.insert(value);
            } else {
                right = new BinaryTreeNode<T>(value);
            }
        }
    }

    public BinaryTreeNode<T> getLeft() {
        return left;
    }

    public BinaryTreeNode<T> getRight() {
        return right;
    }

    public T getValue() {
        return value;
    }

    public boolean isLeaf() {
        return left == null && right == null;
    }
}

File: src/csharp/eu/sig/training/ch03/binarytree/BinaryTreeNode.cs
using System;

namespace eu.sig.training.ch03.binarytree
{

    public class BinaryTreeNode<T> where T : IComparable
    {
        public T Value { get; set; }
        public BinaryTreeNode<T> Left { get; set; }
        public BinaryTreeNode<T> Right { get; set; }

        public BinaryTreeNode(T value)
        {
            this.Value = value;
        }

        public void Insert(T value)
        {
            if (value.CompareTo(this.Value) < 0)
            {
                if (Left != null)
                {
                    Left.Insert(value);
                }
                else
                {
                    Left = new BinaryTreeNode<T>(value);
                }
            }
            else
            {
                if (Right != null)
                {
                    Right.Insert(value);
                }
                else
                {
                    Right = new BinaryTreeNode<T>(value);
                }
            }
        }

        public bool IsLeaf()
        {
            return Left == null && Right == null;
        }
    }
}
